cryptofeed/exchange/okex.py

```diff
@@ -66,51 +69,301 @@ def _entry_is_index_symbol(cls, entry) -> bool:
+                                receipt_timestamp=timestamp
+                                )
+
+    async def _trade(self, msg: dict, timestamp: float):
+        """
+        {
+            "arg": {
+                "channel": "trades",
+                "instId": "BTC-USD-191227"
+            },
+            "data": [
+                {
+                    "instId": "BTC-USD-191227",
+                    "tradeId": "9",
+                    "px": "0.016",
+                    "sz": "50",
+                    "side": "buy",
+                    "ts": "1597026383085"
+                }
+            ]
+        }
+        """
+        for trade in msg['data']:
+            await self.callback(TRADES,
+                                feed=self.id,
+                                symbol=self.exchange_symbol_to_std_symbol(trade['instId']),
+                                order_id=trade['tradeId'],
+                                side=BUY if trade['side'] == 'buy' else SELL,
+                                amount=Decimal(trade['sz']),
+                                price=Decimal(trade['px']),
+                                timestamp=timestamp_normalize(self.id, int(trade['ts'])),
+                                receipt_timestamp=timestamp
+                                )
+
+    async def _funding(self, msg: dict, timestamp: float):
+        for update in msg['data']:
+            # OKEx, OKCoin use the previous period's funding rate (`funding_rate`) at the funding time
+            interval = 60*60*8  # 8 hours
+            funding_time = timestamp_normalize(self.id, int(update['fundingTime']))
+
+            await self.callback(FUNDING,
+                                feed=self.id,
+                                symbol=self.exchange_symbol_to_std_symbol(update['instId']),
+                                timestamp=timestamp,
+                                receipt_timestamp=timestamp,
+                                rate=Decimal(update['fundingRate']),
+                                funding_time=funding_time,
+                                next_funding_rate=Decimal(update['nextFundingRate']),
+                                next_funding_time=funding_time + interval,
+                                interval=interval)
+
+    async def _book(self, msg: dict, timestamp: float):
+        if msg['action'] == 'snapshot':
+            # snapshot
+            pair = self.exchange_symbol_to_std_symbol(msg['arg']['instId'])
+            for update in msg['data']:
+                self.l2_book[pair] = {
+                    BID: sd({
+                        Decimal(price): Decimal(amount) for price, amount, *_ in update['bids']
+                    }),
+                    ASK: sd({
+                        Decimal(price): Decimal(amount) for price, amount, *_ in update['asks']
+                    })
+                }
+                if self.validate_checksum() and self.__calc_checksum(pair) != (update['checksum'] & 0xFFFFFFFF):
```

Comment:
Probably need to change this to happen once every 100 times right?

Reply:
yes, but not for snapshots, only for the updates (below)